File: inkscape/isk-output.rb
# ISK - A web controllable slideshow system
#
# ISK export plugin for inkscape, allows for one-click
# exporting the slide back to the server
#
# Author::    Vesa-Pekka Palmu
# Copyright:: Copyright (c) 2012-2013 Vesa-Pekka Palmu
# License::   Licensed under GPL v3, see LICENSE.md

require "optparse"
require "ostruct"
require "net/http"
require "rexml/document"

options = OpenStruct.new
options.username = nil
options.password = nil
options.slidename = nil

# Per the .inx extension definition we will get the following command-line parameters:
# --username for the ISK login username
# --password for the ISK password
# --iskhost for the ISK address
# last argument given will be the location for a file with the svg data in it.
# inkscape also returns the selected item's svg-id with --id command line parameter
# but we don't care about it

# Parse the command line options
OptionParser.new do |opts|
  opts.banner = "Usage: example.rb [options]"

  opts.on("-u", "--username username", "ISK username") do |user|
    options.username = user
  end

  opts.on("-p", "--password password", "ISK password") do |pw|
    options.password = pw
  end

  opts.on("-n", "--slidename name", "Slide name") do |sn|
    options.slidename = sn
  end

  opts.on("-i", "--iskhost hostname", "ISK hostname") do |host|
    options.host = host
  end

  # Need to specify this even when we don't use the id of selected object
  opts.on("-e", "--id object_id", "Object id") do |id|
    options.id = id
  end
end.parse!

# Send a POST request to ISK and collect cookies
http = Net::HTTP.new(options.host)
resp, data = http.post("/login", "username=#{options.username}&password=#{options.password}")
cookie = resp.response["set-cookie"].split("; ")[0]

#Check the return code from the POST request
unless resp.is_a? Net::HTTPFound
  abort "Error loggin into ISK, aborting"
end

# Store the session cookie
headers = {
  "Cookie" => cookie,
 }

# Read the svg data from the file provided in command line arguments
svg_data = File.read(ARGV.last)

xml = REXML::Document.new svg_data

#Find the first <metadata> element in the svg document
metadata = REXML::XPath.first(xml, "//metadata")
data = metadata.text

# The metadata-element contains a cookie about the origin, it's in the form of slide_id!isk_host
# so for example 123!http://example.com/ we don't care about the legacy url anymore, so just take
# the id
id, isk_server = data.split("!")

# Inkscape chances the background picture from relative to absolute path, so we
# need to change it back. The path will be something like c:/foo/bar/background/empty.png
# and we want /backgrounds/empty, so match on backgrounds and mangle based on that.
REXML::XPath.each(xml, '//image[@id="background_picture"]') do |element|
  href = element.attribute("xlink:href").to_s
  element.add_attribute "xlink:href", "backgrounds#{href.partition("backgrounds").last}"
end

# Write the modified svg into a string
post_svg = String.new
xml.write post_svg

# post the svg data to ISK
resp, data = http.post "/slides/#{id}/svg_data", { "svg" => post_svg }, headers

# Inkscape expects the plugin to return the svg in stdout, so do so
puts svg_data

File: inkscape/isk-output.py
# coding=utf-8

# ISK - A web controllable slideshow system
#
# ISK export plugin for inkscape, allows for one-click
# exporting the slide back to the server
#
# Originally work of Vesa-Pekka Palmu
# This is an Python rewrite of isk-output.rb
#
# Author::    Jarkko Räsänen
# Copyright:: Copyright (c) 2015 Jarkko Räsänen
# License::   Licensed under GPL v3, see LICENSE.md

import urllib
import urllib2
import cookielib
import sys
from xml.dom import minidom
from sys import argv

value = 1
username = ""
password = ""
slidename = ""
hostname = ""
svg_file = ""

while 1:
	if value >= len(argv):
		break
	if argv[value].split("=")[0] == "--username":
		username = argv[value].split("=")[1]
		value = value+1
	
	elif argv[value].split("=")[0] == "--password":
		password = argv[value].split("=")[1]
		value = value+1
		
	elif argv[value].split("=")[0] == "--slidename":
		slidename = argv[value].split("=")[1]
		value = value+1
	
	elif argv[value].split("=")[0] == "--iskhost":
		hostname = argv[value].split("=")[1]
		value = value+1
		
	elif argv[value].split("=")[0] == "--id":
		value = value+1
		
	elif ".svg" in argv[value]:
		svg_file = argv[value]
		value = value+1
	else:
		#print "wat is dis"
		break

try:
	if "username" != "":
		pass
	if "password" != "":
		pass
	if "slidename" != "":
		pass
	if "hostname" != "":
		pass
	if "svg_file" != "":
		pass
	else:
		#print "Some parameter(s) is missing!"
		raise Exception
except:
	#print "Missing mandatory arguments!"
	raise SystemExit

cookiejar = cookielib.CookieJar()
opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cookiejar))

# Login info
payload = {
	"username": username,
	"password": password
}

# Let's make the request ready.
data = urllib.urlencode(payload)
req = urllib2.Request(hostname+"/login", data)

# For login purpose, CookieProcessor saves the cookie to the opener after this.
resp = opener.open(req)

# Open the SVG file for manipulation.
xmldoc = minidom.parse(svg_file)

# Let's get all necessary tags and pieces...
val = xmldoc.getElementsByTagName("image")
id = str(xmldoc.getElementsByTagName("metadata")[0].childNodes[0].toxml().split("!")[0])

# And search the right one which we are going to replace with another one (relative path)
for i in val:
	if "backgrounds" in i.attributes["xlink:href"].value:
		i.attributes["xlink:href"].value = "backgrounds/empty.png"


# Let's get the newly manipulated data, and dump it to ISK server and to STDOUT as required by Inkscape.
newsvg = xmldoc.toxml("UTF-8")

payload = {
	"svg": newsvg
}

data = urllib.urlencode(payload)
req = urllib2.Request(hostname+"/slides/%s/svg_data"%id, data)
resp = opener.open(req)

sys.stdout.write(newsvg)
